File: figs/rtree.lua
package.path = package.path .. ';lib/?.lua'
local luamp = require 'luamp'
local stream = require 'stream'
local table = require 'table'

local tree = luamp.layouts.tree(
    luamp.origin,
    1, 1.2,
    {false,
        {false,
            {false},
            {false}},
        {false,
            {false},
            {false}}})

local root = tree[1]
local left = tree[2]
local right = tree[3]
local left_left = left[2]
local left_right = left[3]
local right_left = right[2]
local right_right = right[3]
left = left[1]
right = right[1]
left_left = left_left[1]
left_right = left_right[1]
right_left = right_left[1]
right_right = right_right[1]

local figs = {}

local function node(center, opts)
    local SEG_WIDTH = 0.5
    local SEG_HEIGHT = 0.4
    table.insert(
        figs,
        luamp.rectangle(center, SEG_WIDTH*2, SEG_HEIGHT, opts))
    table.insert(
        figs,
        luamp.line(
            center + luamp.point(0, SEG_HEIGHT/2),
            center + luamp.point(0, -SEG_HEIGHT/2),
            opts))
    return stream
        .iterate(
            center + luamp.point(-SEG_WIDTH/2, 0),
            function(x)
                return x + luamp.point(SEG_WIDTH, 0)
            end)
        :map(function(x)
            return luamp.rectangle(x, SEG_WIDTH, SEG_HEIGHT)
        end)
        :take(2)
        :collect()
end

local function tree_edge(root_node, left_node, right_node)
    local function draw_edge(seg1, seg2)
        table.insert(
            figs,
            luamp.arrow(
                luamp.centroid(seg1:vertices()[3], seg1:vertices()[4]),
                luamp.centroid(seg2:vertices()[1], seg2:vertices()[2])))
    end
    draw_edge(root_node[1], left_node[1])
    draw_edge(root_node[2], right_node[1])
end

local segs_root = node(root)
local segs_left = node(left, {pen_color=luamp.colors.red})
local segs_right = node(right, {pen_color=luamp.colors.red})
local segs_left_left = node(left_left, {pen_color=luamp.colors.blue})
local segs_left_right = node(left_right, {pen_color=luamp.colors.blue})
local segs_right_left = node(right_left, {pen_color=luamp.colors.blue})
local segs_right_right = node(right_right, {pen_color=luamp.colors.blue})

tree_edge(segs_root, segs_left, segs_right)
tree_edge(segs_left, segs_left_left, segs_left_right)
tree_edge(segs_right, segs_right_left, segs_right_right)

table.insert(figs, luamp.text(segs_left_left[1]:center(), luamp.directions.center, '$a$'))
table.insert(figs, luamp.text(segs_left_left[2]:center(), luamp.directions.center, '$b$'))
table.insert(figs, luamp.text(segs_left_right[1]:center(), luamp.directions.center, '$c$'))
table.insert(figs, luamp.text(segs_left_right[2]:center(), luamp.directions.center, '$d$'))
table.insert(figs, luamp.text(segs_right_left[1]:center(), luamp.directions.center, '$e$'))
table.insert(figs, luamp.text(segs_right_left[2]:center(), luamp.directions.center, '$f$'))
table.insert(figs, luamp.text(segs_right_right[1]:center(), luamp.directions.center, '$g$'))
table.insert(figs, luamp.text(segs_right_right[2]:center(), luamp.directions.center, '$h$'))

print(luamp.figure(table.unpack(figs)))

File: figs/rtree.py
from fathom import Point, ORIGIN
from fathom.tikz import Canvas
from fathom.geometry import *
import fathom.layout as layout
import fathom.tikz.colors as colors
import fathom.tikz.line_styles as line_styles
import fathom.tikz.locations as locations
from itertools import *

def node(canvas, center, **kws):
    SEG_WIDTH = 0.5
    SEG_HEIGHT = 0.4

    vs = repeat(center + Point(-SEG_WIDTH / 2, 0))
    vs = accumulate(vs, lambda x, _: x + Point(SEG_WIDTH, 0))
    vs = islice(vs, 2)
    vs = [Rectangle(center=x, width=SEG_WIDTH, height=SEG_HEIGHT)
        for x in vs]

    canvas.new_rectangle(
        center=center,
        width=SEG_WIDTH * 2,
        height=SEG_HEIGHT,
        **kws)
    canvas.new_line(
        src=center + Point(0, SEG_HEIGHT / 2),
        dst=center + Point(0, -SEG_HEIGHT / 2),
        **kws)

    return vs

def tree_edge(canvas, root, left, right):
    def draw_edge(seg0, seg1):
        canvas.new_arrow(
            src=centroid([seg0.vertices()[2], seg0.vertices()[3]]),
            dst=centroid([seg1.vertices()[0], seg1.vertices()[1]]))
    draw_edge(root[0], left[0])
    draw_edge(root[1], right[0])

if __name__ == '__main__':
    tree = layout.tree(
        ['root',
         ['left', ['left_left'], ['left_right']],
         ['right', ['right_left'], ['right_right']]],
        root=ORIGIN,
        h_sep=1.2,
        v_sep=1)

    canvas = Canvas(
        preamble=['\\usepackage{amsmath}'],
        leading_instructions=['\\footnotesize'])

    segs_root = node(canvas, tree['root'])
    segs_left = node(canvas, tree['left'], pen_color=colors.RED)
    segs_right = node(canvas, tree['right'], pen_color=colors.RED)
    segs_left_left = node(canvas, tree['left_left'], pen_color=colors.BLUE)
    segs_left_right = node(canvas, tree['left_right'], pen_color=colors.BLUE)
    segs_right_left = node(canvas, tree['right_left'], pen_color=colors.BLUE)
    segs_right_right = node(canvas, tree['right_right'], pen_color=colors.BLUE)

    tree_edge(canvas, segs_root, segs_left, segs_right)
    tree_edge(canvas, segs_left, segs_left_left, segs_left_right)
    tree_edge(canvas, segs_right, segs_right_left, segs_right_right)

    segs = segs_left_left + segs_left_right + segs_right_left + segs_right_right
    ts = ['$a$', '$b$', '$c$', '$d$', '$e$', '$f$', '$g$', '$h$']
    for r, t in zip(segs, ts):
        canvas.new_text(anchor=r.center(), text=t)

    print(canvas.draw())
